src/hooks/useWatchHistory.ts

```diff
@@ -0,0 +1,123 @@
+ * problem is that some events are triggered when the `item` to update has changed. For example, when clicking a media
+ * item in the "Related shelf". This causes the wrong item to be saved in the watch history.
+ *
+ * __The solution:__
+ *
+ * This hook listens to the player time update event and queues a watch history entry with the current progress and
+ * item. When this needs to be saved, the queue is used to look up the last progress and item and save it when
+ * necessary. The queue is then cleared to prevent duplicate saves and API calls.
+ *
+ * @param player
+ * @param item
+ */
+export const useWatchHistory = (player: JWPlayer | undefined, item: PlaylistItem) => {
+  const queuedWatchProgress = useRef<QueuedProgress | null>(null) as MutableRefObject<QueuedProgress | null>;
+
+  // config
+  const { features } = useConfigStore((s) => s.config);
+  const continueWatchingList = features?.continueWatchingList;
+  const watchHistoryEnabled = !!continueWatchingList;
+
+  // watch History
+  const watchHistoryItem = useWatchHistoryStore((state) => (!!item && watchHistoryEnabled ? state.getItem(item) : undefined));
+
+  // update the queued watch progress on each time update event
+  const handleTimeUpdate = useEventCallback(() => {
+    if (!player) return;
+
+    const progress = player.getPosition() / item.duration;
+
+    if (!isNaN(progress) && isFinite(progress)) {
+      queuedWatchProgress.current = {
+        item,
+        progress,
+      };
+    }
+  });
+
+  // maybe store the watch progress when we have a queued watch progress
+  const maybeSaveWatchProgress = useCallback(() => {
+    if (!watchHistoryEnabled || !queuedWatchProgress.current) return;
+
+    const { item, progress } = queuedWatchProgress.current;
+
+    // save the queued watch progress
+    saveItem(item, progress);
+
+    // clear the queue
+    queuedWatchProgress.current = null;
+  }, [watchHistoryEnabled]);
+
+  // listen for certain player events
+  useEffect(() => {
+    if (!player || !watchHistoryEnabled) return;
+
+    player.on('time', handleTimeUpdate);
+    player.on('pause', maybeSaveWatchProgress);
+    player.on('complete', maybeSaveWatchProgress);
+    player.on('remove', maybeSaveWatchProgress);
+
+    return () => {
+      player.off('time', handleTimeUpdate);
+      player.off('pause', maybeSaveWatchProgress);
+      player.off('complete', maybeSaveWatchProgress);
+      player.off('remove', maybeSaveWatchProgress);
+    };
+  }, [player, watchHistoryEnabled, maybeSaveWatchProgress, handleTimeUpdate]);
+
+  useEffect(() => {
+    return () => {
+      // store watch progress on unmount and when the media item changes
+      maybeSaveWatchProgress();
+    };
+  }, [item?.mediaid, maybeSaveWatchProgress]);
+
+  // add event listeners for unload and visibility change to ensure the latest watch progress is saved
+  useLayoutEffect(() => {
```

Comment:
From what I can tell, onunload and onvisibilitychange are full of small differences and bugs between browsers, especially on older and mobile versions.

Based on these 2 pages:
- https://www.igvita.com/2015/11/20/dont-lose-user-and-app-state-use-page-visibility/
- https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event

it looks like we might also want to include the pagehide event for fullest coverage. How would you feel about also saving the progress on a timeout every couple minutes or so? Even if it's only every 5 minutes, it would greatly reduce our risk of losing progress for longer content.

Reply:
Yeah, no problem! I've added the `pagehide` event listener and a progressive save function.

Reply:
I found another bug. Due to the `watchHistoryController#saveItem` method always calling the `getSeriesByMediaIds`, all page hide / visibility change events are lost. This is because the XHR request is canceled (or never fulfilled) and the localStorage is never updated.

But I also realize that when using Cleeng/JWP the save call might also be skipped in this scenario...

Reply:
We could drill the `seriesId`; `MediaSeries` -> `InlinePlayer`/`Cinema` -> `PlayerContainer`

Reply:
Perhaps we shouldn't be relying on these events at all? The progress is now saved when pausing and progressively every 5 minutes. 

We should be using the `sendBeacon` API for beforeunload/visibilitychange when the user closes the tab or browser to reliably save the progress. But this requires an API that can handle all scenarios and doesn't work for localStorage.

When I remove the `getSeriesByMediaIds` call, it does work when saving the progress to localStorage. But I don't think it works when using Cleeng/JWP.

Reply:
I haven't dug deep into this, just a quick google, but maybe we can add keep-alive to the api call?
https://stackoverflow.com/questions/60530716/execute-server-api-call-on-beforeunload-limitations-on-page-unload

Otherwise, periodic updates is fine with me, but I'd make them closer to every minute instead if it's all we have. Or maybe even just fire off one every time we get a time event with some logic to keep them from backing up?

Reply:
Neat! It should be. We've made some other small tweaks to the SDK, but I'll have to check with the team.